### PR TITLE
Add S3 SSE-C support

### PR DESCRIPTION
**Pull request by xangelix**
Adds support for [SSE-C](https://docs.aws.amazon.com/AmazonS3/latest/userguide/ServerSideEncryptionCustomerKeys.html).

Needs testing, not ready to be merged.

### CODE REVIEW

**Comment by xangelix** [comment]
Appears to now be working well--appreciate anyone else who's able to give it a go.

**Comment by clokep** [comment]
@xangelix Looks like there's some linting and test issues!

**Review by clokep — CHANGES_REQUESTED** [review]
Marking as needing changes due to the failing build.

**Comment by DMRobertson** [comment]
See #83 for the integration test.

**Comment by xangelix** [comment]
Thanks for the quick response, linting errors are now fixed.

> See #83 for the integration test.

After the changes here, I am unable to reproduce any failing tests unique to my branch.
However, I was also unable to reproduce a passing set of builds on main:

#40c6a5 (latest main)
https://github.com/xangelix/synapse-s3-storage-provider/actions/runs/3357238074

#8383d2 (latest this pr)
https://github.com/xangelix/synapse-s3-storage-provider/actions/runs/3357254105

This is my first time dealing with GitHub actions, so I may be overlooking something, but it does not appear to me the failing test on 3.7 has to do with this pull request?
I'm not certain how the passing build at https://github.com/matrix-org/synapse-s3-storage-provider/actions/runs/3320569047/jobs/5487117200 (on #40c6a5) was accomplished.

File: README.md [hunk 1]
```diff
@@ -28,6 +28,11 @@ media_storage_providers:
     access_key_id: <S3_ACCESS_KEY_ID>
     secret_access_key: <S3_SECRET_ACCESS_KEY>
 
+    # Server Side Encryption for Customer-provided keys
+    #sse_customer_key: <S3_SSEC_KEY>
+    # Your SSE-C algorithm is very likely AES256
```
Comment (reivilibre) [inline README.md:33]:
```suggestion
    # Your SSE-C algorithm is very likely AES256
    # Default is AES256.
```

File: scripts/s3_media_upload [hunk 2]
```diff
@@ -495,6 +501,17 @@ def main():
...
     )
 
+    upload_parser.add_argument(
+        "--sse-customer-key", help="SSE-C key to use",
+    )
+
+    upload_parser.add_argument(
+        "--sse-customer-algo",
+        help="Algorithm for SSE-C, only used if sse-customer-key is also specified",
+        nargs="?",
```
Comment (reivilibre) [inline scripts/s3_media_upload:511]: I don't think `nargs=?` is what we want here: that lets you specify the option without giving it a value. c.f. https://docs.python.org/3/library/argparse.html#nargs

```
>>> import argparse
>>> p = argparse.ArgumentParser()
>>> p.add_argument("--foo", nargs="?", default="lol")
_StoreAction(option_strings=['--foo'], dest='foo', nargs='?', const=None, default='lol', type=None, choices=None, required=False, help=None, metavar=None)
>>> p.parse_args(["--foo"])
Namespace(foo=None)
>>> p.parse_args([])
Namespace(foo='lol')
>>> p = argparse.ArgumentParser()
>>> p.add_argument("--foo", default="lol")
_StoreAction(option_strings=['--foo'], dest='foo', nargs=None, const=None, default='lol', type=None, choices=None, required=False, help=None, metavar=None)
>>> p.parse_args([])
Namespace(foo='lol')
>>> p.parse_args(["--foo"])
usage: [-h] [--foo FOO]
: error: argument --foo: expected one argument
```

```suggestion
```

**Review by reivilibre — CHANGES_REQUESTED** [review]
This looks reasonable to me but I obviously haven't tested it, nor are there any additional tests here. (Not sure whether this is a reasonable thing to add tests for though; I feel like they'd be very superficial.)

File: s3_storage_provider.py [hunk 3]
```diff
@@ -62,7 +62,7 @@ class S3StorageProviderBackend(StorageProvider):
     def __init__(self, hs, config):
         self.cache_directory = hs.config.media.media_store_path
         self.bucket = config["bucket"]
-        self.storage_class = config["storage_class"]
+        self.eargs = config["eargs"]
```
Comment (squahtx) [inline s3_storage_provider.py:65]:
```suggestion
        # A dictionary of extra arguments for uploading files.
        # See https://boto3.amazonaws.com/v1/documentation/api/latest/reference/customizations/s3.html#boto3.s3.transfer.S3Transfer.ALLOWED_UPLOAD_ARGS
        # for a list of possible keys.
        self.eargs = config["eargs"]
```

Comment (squahtx) [inline s3_storage_provider.py:65]: Can we unabbreviate `eargs` to `extra_args` please?
It's not clear what the `e` stands for unless you stumble upon the call to `upload_file`.

**Review by reivilibre — APPROVED** [review]

**Pull request merged**